compiler/ballerina-lang/src/main/java/io/ballerina/projects/Package.java

```diff
@@ -578,34 +578,58 @@ private Package createNewPackage() {
+                    deleteCaches(dependency, oldGraph);
+                }
+            }
+            diff = newGraph.difference(oldGraph);
+            if (!diff.isEmpty()) {
+                // A non-empty diff means there can be dependant nodes in the old graph that
+                // need to be recompiled. Hence we remove the dependant modules from the package cache.
                 for (ResolvedPackageDependency dependency : diff) {
-                    for (ModuleId moduleId : dependency.packageInstance().moduleIds()) {
-                        if (!dependency.packageInstance().descriptor().isLangLibPackage()) {
-                            Module module = dependency.packageInstance().module(moduleId);
-                            PackageID packageID = module.descriptor().moduleCompilationId();
-                            // remove the module from the compiler packageCache
-                            packageCache.remove(packageID);
-                            // reset the module in the project environment packageCache to make the module recompile
-                            // and add symbols
-                            environmentPackageCache.removePackage(module.moduleId().packageId());
-                            module.moduleContext().setCompilationState(null);
-                        }
+                    for (ResolvedPackageDependency directDependent : newGraph.getDirectDependents(dependency)) {
+                        deleteCaches(directDependent, newGraph);
                     }
                 }
             }
-            return this.project.currentPackage();
+        }
+
+        private void deleteCaches(ResolvedPackageDependency dependency,
+                                  DependencyGraph<ResolvedPackageDependency> graph) {
+            if (dependency.equals(graph.getRoot())) {
+                return;
+            }
+            CompilerContext compilerContext = project.projectEnvironmentContext()
+                    .getService(CompilerContext.class);
+            PackageCache packageCache = PackageCache.getInstance(compilerContext);
+            for (ModuleId moduleId : dependency.packageInstance().moduleIds()) {
+                if (!dependency.packageInstance().descriptor().isLangLibPackage()) {
+                    Module module = dependency.packageInstance().module(moduleId);
+                    PackageID packageID = module.descriptor().moduleCompilationId();
+                    // remove the module from the compiler packageCache
+                    packageCache.remove(packageID);
+                    // reset the module in the project environment packageCache to make the module recompile
+                    // and add symbols
+                    module.moduleContext().setCompilationState(null);
+                }
+            }
+            for (ResolvedPackageDependency directDependent : graph.getDirectDependents(dependency)) {
```

Comment:
We can further optimize this logic by introducing `getAllDependents` method to the dependency graph.

https://github.com/ballerina-platform/ballerina-lang/issues/36113